docs/02-app/01-building-your-application/09-authentication/index.mdx

```diff
@@ -0,0 +1,345 @@
+
+Here's how to implement middleware for authentication in Next.js:
+
+1. **Setting Up Middleware:**
+   - Create a `middleware.ts` or `.js` file in your project's root directory.
+   - Include logic to authorize user access, such as checking for authentication tokens.
+2. **Defining Protected Routes:**
+   - Use the `matcher` option in your middleware to identify routes that require authorization checks.
+3. **Middleware Logic:**
+   - Write logic to verify if a user is authenticated. Check user roles or permissions for route authorization.
+4. **Handling Unauthorized Access:**
+   - Redirect unauthorized users to a login or error page as appropriate.
+
+Example middleware file:
+
+```ts filename="middleware.ts" switcher
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  const currentUser = request.cookies.get('currentUser')?.value
+
+  if (currentUser) {
+    return Response.redirect(new URL('/dashboard', request.url))
+  }
+  return Response.redirect(new URL('/login', request.url))
+}
+
+export const config = {
+  matcher: '/dashboard',
+}
+```
+
+This example uses [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect_static) for efficient, centralized access control. It handles redirects early in the request pipeline, making it efficient and ideal for broad strategies.
+
+For specific redirection needs, the `redirect` function can be used in Server Components, Route Handlers, and Server Actions to provide more control. This is useful for role-based navigation or context-sensitive scenarios.
+
+```js
+import { redirect } from 'next/navigation'
+
+export default function Page() {
+  // Logic to determine if a redirect is needed
+  const accessDenied = true
+  if (accessDenied) {
+    redirect('/login')
+  }
+
+  // Define other routes and logic
+}
+```
+
+This flexibility is key when handling redirections based on detailed authentication outcomes or other contextual factors. For example, `redirect('/login')` in a Server Action allows for precise redirection based on the user's role or specific conditions during the authentication process.
+
+### Protecting Server Actions
+
+It is important to treat [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) with the same security considerations as public-facing API endpoints. Verifying user authorization for each action is crucial. Implement checks within Server Actions to determine user permissions, such as restricting certain actions to admin users.
+
+In the example below, we check the user's role before allowing the action to proceed:
+
+```ts filename="app/lib/actions.ts" switcher
+'use server'
+
+// ...
+
+export async function serverAction(request) {
```

Comment:
- The `request` argument is not being used in the function, you can delete it. 
- Consider calling the action something that reflects a real-world example instead of `serverAction`.

```suggestion
export async function serverAction() {
```